components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/util/APIClient.java

```diff
@@ -147,25 +153,36 @@ private ActionInvocationResponse handleResponse(HttpResponse response) {
             case HttpStatus.SC_BAD_GATEWAY:
             case HttpStatus.SC_SERVICE_UNAVAILABLE:
             case HttpStatus.SC_GATEWAY_TIMEOUT:
-                actionInvocationResponseBuilder.errorLog(
-                        "Failed to execute the action request. Received status code: " + statusCode + ".");
+                handleServerError(actionInvocationResponseBuilder, responseEntity, statusCode);
                 actionInvocationResponseBuilder.retry(true);
                 break;
             default:
-                actionInvocationResponseBuilder.errorLog("Unexpected response received with status code: " + statusCode
-                        + ".");
+                handleError(actionInvocationResponseBuilder, responseEntity, statusCode);
                 break;
         }
 
         return actionInvocationResponseBuilder.build();
     }
 
-    private void handleSuccess(ActionInvocationResponse.Builder builder, HttpEntity entity, int statusCode) {
+    private void handleSuccessOrFailure(ActionInvocationResponse.Builder builder, HttpEntity entity, int statusCode) {
 
         try {
-            builder.response(handleSuccessResponse(entity));
+            builder.response(handleSuccessOrFailureResponse(entity));
         } catch (ActionInvocationException e) {
             builder.errorLog("Unexpected response for status code: " + statusCode + ". " + e.getMessage());
+            ActionInvocationErrorResponse invalidResponse = handleInvalidResponse();
+            builder.response(invalidResponse);
+        }
+    }
+
+    private void handleError(ActionInvocationResponse.Builder builder, HttpEntity entity, int statusCode) {
+
+        try {
+            builder.response(handleErrorResponse(entity));
+        } catch (ActionInvocationException e) {
+            builder.errorLog("Unexpected response for status code: " + statusCode + ". " + e.getMessage());
+            ActionInvocationErrorResponse invalidResponse = handleInvalidResponse();
+            builder.response(invalidResponse);
```

Comment:
what exactly is the reason to set this error response. This is not coming from the external service. The use of 'ActionInvocationErrorResponse' is to process the error responses coming from external service

Reply:
This comment applies for all similar cases done below

Reply:
This has been utilized to communicate the following exception scenarios from the token endpoint as a `server error (500)`.

- When an unexpected JSON response was received for 200 status code (SUCCESS or FAILED status).
- When unexpected JSON response was received for ERROR status (both client errors and server errors).

Reply:
This has been refactored.